### ISSUE TITLE
GSoC 2015 add-on system proposals review

### ISSUE THREAD

**Issue by ysitu** [networkx/networkx#1429]
@SanketDG @OrkoHunter @hagberg @dschult @jtorrents @chebee7i @Midnighter

## Before we start
The review below is based on the proposal as they appear in the [wiki](https://github.com/networkx/networkx/wiki) on Mar 23, 2015.

## Summary of proposals
There are currently two proposals regarding creating an add-on system for NetworkX, [one](https://github.com/networkx/networkx/wiki/Add-on-system-for-NetworkX:-GSOC-2015-Application-Sanket-Dasgupta) from @SanketDG and [one](https://github.com/networkx/networkx/wiki/GSoC-2015-Application-Himanshu-Mishra-:-Add-On-System-for-Networkx) from @OrkoHunter. The core part of @SanketDG's proposal is relatively sketchy, largely paraphrasing what has been mentioned in the [project ideas page](https://github.com/networkx/networkx/wiki/GSoC-2015-project-ideas). In comparison, @OrkoHunter's proposal goes into more details of the specifics of execution plan and timeline.

## Initial review (project idea-only)
The minimum requirement of an add-on system is to allow installation of separately distributed packages into the `networkx.external.addons` namespace. Both proposals address this need. `setuptools ` has largely provided [support](https://pythonhosted.org/setuptools/setuptools.html#namespace-packages) for this use case. So if we are to do this alone, the "system" part of "add-on system" fades away, and creation of various add-ons for NetworkX takes up most weight. This readily covers all that is mentioned in the project ideas page. If executed well, this *will*  make for a meaningful GSoC project.

However, if we are not to let `setuptools` claim most if not all the fame in enabling an add-on system, we can explore what it does not already do. For example, we can investigate tighter integration with the NetworkX core than mere out-of-tree module distribution. Consider incorporation of network algorithms from LEMON for instance. Suppose that we want to use LEMON to provide an implementation of the network simplex method instead of `networkx.algorithms.flow.network_simplex`. Should `networkx.algorithms.flow.network_simplex` have hard-coded knowledge about `networkx.external.addons.lemon`? Or should such knowledge be supplied via dependency injection? Or should it provide a `flow_func` interface a la `networkx.algorithms.flow;maximum_flow`? Or should we simply let `networkx.external.addons.lemon` nuke and replace `networkx.algorithms.flow.network_simplex` at load-time without the latter's knowning? How does this one particular instance generalize to other parts of NetworkX? Can we standardize the approach? Thoughts and decisions on trade-off raised by these and similar questions will make proposals and projects richer in nature. If we pursue this route or some alternative similar in spirit, we put more weight in developing a new ecosystem for NetworkX (and may accordingly reduce the workload of creating the add-ons themselves), which I think will have greater long-term impact.

**Comment by OrkoHunter** [comment]
Concerning `networkx.algorithms`, suppose a person wants to make use of a third-party library to implement a function, then it's sure that they would like to use the pre-existing algorithm of NetworkX as a second choice. So, if someone has installed the concerned `networkx-<add-on>`, they would by default, expect `networkx.algorithms.<the-function>` to give output with help of the add-on. However, in some situations, they can still want to use a function in the NetworkX way. So, what *we* can do to is to create a "system" in which

* If the add-on is not installed, `networkx.algorithms` returns outputs according to the NetworkX algorithms.

* As the add-on gets installed, `networkx.algorithms` starts using the add-on's algorithms by default. 

* Nonetheless, there will still exist a way or I should say a switch, which will allow to decide whether we want to use the add-on's method or the NetworkX proper method. Uninstalling should be least preferred.

I think this could be a good approach. Suggestions?

**Comment by SanketDG** [comment]
Thank you for letting me know about this.

Unfortunately, I wont be taking part in GSoc 2015 as I am having difficulties with my university enrollment form. I believe @OrkoHunter will implement a very good system and I wish him luck.

But I am willing to contribute to this project. In a non-GSoC way. Thanks for the detailed information. I am now clear about the project objectives.

**Comment by OrkoHunter** [comment]
> Nonetheless, there will still exist a way or I should say a switch, which will allow to decide whether we want to use the add-on's method or the NetworkX proper method. Uninstalling should be least preferred.

We can just pass an argument `use_addon=False`

**Comment by OrkoHunter** [comment]
@SanketDG 

> Unfortunately, I wont be taking part in GSoc 2015 as I am having difficulties with my university enrollment form.

This is really sad. Did you submit the enrollment form and it got rejected? If you are in college, you can resolve this issue legitimately some how.

>I believe @OrkoHunter will implement a very good system and I wish him luck.

Thank you.

> But I am willing to contribute to this project. In a non-GSoC way. Thanks for the detailed information. I am now clear about the project objectives.

That's really great to hear!

**Comment by jtorrents** [comment]
I did not think much about this, but I'd prefer to be very explicit about add-ons functionality. This will force users to think in advance how to use the addons in their code. We could require to import the addon and use that namespace to call add-on functions. Something like:

```python
import networkx as nx
from networkx.external.addons import lemon

G = build_a_graph_with_appropiate_attrs()
result_nx = nx.network_simplex(G, demand='demand', capacity='capacity', weight='weight')
result_lemon = lemon.network_simplex(G, demand='demand', capacity='capacity', weight='weight')
assert result_nx == result_lemon
```

This will reduce the 'magic' for the user and the development effort on NetworkX side. I think the one of the main points of having addons is to allow us to use NetworkX API for interacting with other software (as much as possible, eg an addon could accept arguments that will be ignored if not applicable in some functions, ...) and also return results following NetworkX conventions. I even think that, for the users, the standarization of the results returned will be the most useful feature. I, as a user, would really love to be able to get a nested pair of dictionaries with the node connectivity between all pairs of nodes really fast :)

**Comment by jtorrents** [comment]
@SanketDG, sorry to hear that. You are very welcomed to contribute anyway.

**Comment by OrkoHunter** [comment]
@jtorrents

> ```
import networkx as nx
from networkx.external.addons import lemon
G = build_a_graph_with_appropiate_attrs()
result_nx = nx.network_simplex(G, demand='demand', capacity='capacity', weight='weight')
result_lemon = lemon.network_simplex(G, demand='demand', capacity='capacity', weight='weight')
assert result_nx == result_lemon
```

This seems to be nice plan. We've been planning to store bindings in a different package, say `networkx-lemon`. We can achieve the above API by putting something like this in the `networkx.external.addons.lemon.py`:

```
try:
    from networkx-lemon import *
except ImportError:
    print("Please install networkx-lemon to use this add-on")  # Or maybe raise an issue?
```

> I even think that, for the users, the standarization of the results returned will be the most useful feature.

Agree. I'll put my efforts to achieve the same.

**Comment by bjedwards** [comment]
What about keeping the implementations separate as @jtorrents suggests, but also provide a method (from networkx from the add-on?) to blow away the underlying implementation. Something like

```python
import networkx as nx
from networkx.external.addons import lemon

G = build_a_graph_with_appropiate_attrs()
# Uses Networkx
result_nx = nx.network_simplex(G, demand='demand', capacity='capacity', weight='weight')
# Uses Lemon
result_lemon = lemon.network_simplex(G, demand='demand', capacity='capacity', weight='weight')

nx.use_addon(lemon.some_function_or_data_structure)
# Now the uses Lemon
result = nx.network_simplex(G, demand='demand', capacity='capacity', weight='weight')
```

This might be more convenient if I have a large codebase and want to replace all instances of the networkx algorithm with some other add-on library.

I don't know how hard this would be, and how much machinery could be taken care of by networkx, and how much might need to be supplied by the add-on developer.

**Comment by OrkoHunter** [comment]
> This might be more convenient if I have a large codebase and want to replace all instances of the networkx algorithm with some other add-on library.

I like it. This will indeed make the transition easier for people with large codebase. I vote for keeping both APIs.

> I don't know how hard this would be, and how much machinery could be taken care of by networkx, and how much might need to be supplied by the add-on developer.

This will surely add some more work e.g. I'll have to tell all these functions to check for a boolean or something to determine what implementation to use.

@ysitu How do you like this approach?

**Comment by OrkoHunter** [comment]
@ysitu I have pushed my final draft to wiki. If there's still any change you want, please post. The deadline is in few hours.

**Comment by ysitu** [comment]
@OrkoHunter Overall I am okay with the contents. We can continue to refine even after the deadline. That said, I want to say that your timeline looks aggressive to me. (I expect the add-on system proper and LEMON integration to take more than what you propose. Of course, you can use the pre-GSoC time to work on them.) But as long as you can keep up, I would not oppose in principle.

**Comment by OrkoHunter** [comment]
@ysitu What I have done in my timeline related to this project was just a small testing of external metis library fetched from your commit. Now, I am doing the work locally and have stopped pushing it over my repository because that makes no sense.

I too want the add-on system proper to be created at its best. For this, we'll have to keep discussing all the aspects openly. This will keep up the spirit to design the best integration possible.

**Comment by MridulS** [comment]
@ysitu @hagberg @dschult @jtorrents @chebee7i @midnighter 5 hours to go for deadline. I would be grateful to know some pointers and views for my application https://github.com/networkx/networkx/wiki/NetworkX-GSoC-2015-Application:-NetworkX-2.0-API---Mridul-Seth

**Comment by ysitu** [comment]
@MridulS I would say no sweat. Like I said to @OrkoHunter, "[we] can continue to refine even after the deadline."

Regarding the contents, I like the way you partition the work into two halves. The first half has received much discussion and is something that we definitely want even without participating in GSoC. The second half is more exploratory and open-ended. But that is okay. I would consider that every proposal just has to have some elements of this nature so that the project does not become some "do what was told to do" kind of thing. Overall your proposal sounds like a workable plan to me.

PS: Purely editorial suggestion: The "Description" section starts with a very long paragraph that is hard to follow. You should break it into several shorter paragraphs.

**Comment by ysitu** [comment]
@OrkoHunter In order to for function replacement by add-ons to work, we need to fix the practice of liberally importing functions into numerous different namespaces. Modules instead of individual functions should instead be imported. Otherwise, an add-on needs to look for all places where a function is imported and override all those references, or else people will be surprised to see different versions of the function being called in different occasions. This clean-up work can be combined with the work on the 2.0 API.

**Comment by OrkoHunter** [comment]
@ysitu I guess you are taking the situation when people would like to replace the existing functions with the add-ons' i.e., use the `nx.use_addon()` as suggested by @bjedwards. I think there's not much problem in practicing the import of modules instead of individual functions. However, I'm still not sure where to start that from and for example what could be some possible changes.

**Comment by ysitu** [comment]
@OrkoHunter That is nontrivial. Think of how many times you have used `nx.shortest_path_length` instead of first importing `networkx.path_to_ssp_module.that_i_never_bothered_to_remember`.

**Comment by OrkoHunter** [comment]
@ysitu Yea, never done the second one.

**Comment by OrkoHunter** [comment]
So not all the functions should be imported through the module and some should be imported in the local namespace as well like the `nx.shortest_path_length`. Then we'll have to decide on the ones which are needed to be fixed accordingly.

**Comment by Midnighter** [comment]
My preferred usage (where I have control over the code) would be the explicit import of an add-on. I really dislike overwriting existing functions (hi R). As you mention it is difficult to ensure that you shadow all networkx versions of a function and it is the opposite of explicit (imho).

What I sometimes do is an alternative import, for example,

```python
try:
    import simplejson as json
except ImportError:
    import json
```

where those modules provide the exact same functionality. Of course, here the problem is that networkx has everything in the top level module, otherwise this could be as simple as:

```python
try:
    import networkx.addon as util
except ImportError:
    import networkx.util as util

result = util.some_function
```

Just my $ 0.02.

**Comment by bjedwards** [comment]
I agree with @Midnighter that silently replacing functions is not my preferred usage. Is it possible that `nx.use_addon(lemon.some_function_or_data_structure)` could replace the relevant functions, then reload networkx in such a way that we get the correct behavior in functions that use the replaced functions? Maybe monkey patching like that isn't a good idea...

**Comment by OrkoHunter** [comment]
@ysitu It was fun for me giving Cython some quality time last week. The large enhancement factor of speed is something really fascinating. I was wondering is it good to Cythonize some of our core algorithms too? I remember #1051 was closed because it had used scipy/numpy. So, I think it'd be a no for Cython too.
I don't know much of the core algorithms, so if there is any part where speed does matter, I can take on that.

**Comment by ysitu** [comment]
@OrkoHunter It would be interesting to detect availability of Cython and Cythonize the *entire* library at install time. But for the start, you can try out the algorithms in `networkx.algorithms.flow`. You can use the networks used in their tests as input.

**Comment by OrkoHunter** [comment]
Hi!
My exams are over and I'll be travelling this week. I can start my work from May 5.
I wrote a mail to @Midnighter @dschult. And @ysitu I don't have any of your contact details. We have to schedule weekly meeting somewhere. Maybe Gitter or E-Mail. You have my contact details.

**Comment by OrkoHunter** [comment]
@bjedwards @Midnighter 
Agreed with you. `nx.use_addon()` should also be accompanied with `some_special_function` as argument.

**Comment by MridulS** [comment]
@ysitu Can you email me your contact info :)

**Comment by OrkoHunter** [comment]
We are going to have our first addon [networkx-metis](https://github.com/networkx-metis) in the networkx umbrella. It was written a year back by @ysitu which implements very fast graph partitioning using METIS.

As the GSoC coding period begins, I'd like to discuss here something which has been taken upon many times but not conclusively.

There are two ways in which the upcoming model can be imagined, which are explicitly different as I mention both of them here altogether.

1) Take a big project for example which depends on it eminently and has a big bundle of code written using networkx. How and why they would like to use any addon or such? 
*Considering 'why'* : Speed is something which surely python makes us compromise with. Upon bench-marking `LEMON` against all co-existing graph packages, they claim to be the fastest ([Page 72 on their intro manual](http://lemon.cs.elte.hu/pub/doc/lemon-intro-presentation.pdf)). `networkx-lemon` is surely much needed here.

*Considering 'how'* : They would like to install an add-on of their need which would make their existing code automatically run faster than before without making changes into their entire library.

2) Second model is a different one, making networkx core feel more secure. It would suffice to be merely able to inject a new module into the networkx namespace. This is a much simpler model which is in need for addons like `networkx-metis`. But we can surely sense that both models have different spirit of their own.

--    
As it seems, the simpler model is going to be implemented first. But we surely need to take a decision about what we have to do about the first one which is important because of many reasons.

@ysitu Correct me if I've said something wrong or missed something.

**Comment by ysitu** [comment]
@OrkoHunter You are correct that we want to do 2) first because it is simpler, perhaps much simpler, than 1). With some minor tweaks in Networkx proper and some setuptools magic, `networkx-metis` should be ready to go. It is essentially implementing separate packages except that those packages are installed into NetworkX's namespace.

1) is much more challenging. Add-ons need a reliable way to redirect all references to predefined functions in NetworkX to alternative implementations much like how mocks in testing work. This is tricky because the existing practices of liberally importing functions, instead of modules, in NetworkX. This creates multiple references to the same function scattered in numerous modules. Properly fixing this requires refactoring to remove all function imports (which itself is actually of great value) and implement a [hack](https://mail.python.org/pipermail/python-ideas/2012-May/014969.html) (recommended by Guido van Rossum) so that top-level imports like `nx.shortest_path` continue to work (backward compatibility is very important in this particular case).

Alternatively, NetworkX can provide intrusive injection points that add-ons can hook onto, which work along the line of an `@overridable` decorator that looks for alternative implementations in add-ons and replace the decorated functions as appropriate. The refactoring effort is less intense as we only need to add decorators to all functions, but the end result is uglier and creates a process problem---every time we receive a new functionality PR in the future, we will need to watch out for the missed decorators---we have not always been successful in getting people to update `api_x.y.rst` in their PRs.

**Comment by MridulS** [comment]
I think this can be closed now.
